### PR TITLE
Lifinity integration part 2

### PR DESCRIPTION
**Pull request by sjillen**
- NFT holder must now be the Governance (via an ATA)
-  got rid of Lifinity lib folder container compiled code provided by Lifinity devs

### CODE REVIEW

File: components/instructions/programs/lifinity.tsx [hunk 1]
```diff
@@ -74,9 +77,12 @@ export const LIFINITY_PROGRAM_INSTRUCTIONS = {
           .shiftedBy(-tokenLPInfo.account.decimals)
           .toString();
 
+        const poolLabel = getPoolLabelByPoolMint(tokenLPAcc.toBase58());
+
         return (
           <>
-            <p>{`Max Amount of Token A to deposit: ${uiAmountTokenA}`}</p>
+            <p>{`Liquidity Pool: ${poolLabel}`}</p>
```
Comment (Orelsanpls) [inline components/instructions/programs/lifinity.tsx:84]: nit: 

We gotta set a standard about how to display data on `instructions/program`  :)

File: pages/dao/[symbol]/proposal/components/instructions/Lifinity/DepositToPool.tsx [hunk 2]
```diff
@@ -61,14 +57,22 @@ const DepositToPool = ({
       form,
       governedAccountPubkey,
     }) {
+      // let's recalculate at the last moment to get the LP amount.
+      const depositAmountOut = await getDepositOut({
+        connection: connection,
+        wallet,
+        amountTokenA: form.amountTokenA!,
```
Comment (Orelsanpls) [inline pages/dao/[symbol]/proposal/components/instructions/Lifinity/DepositToPool.tsx]: Is it ui amount?

File: tools/sdk/lifinity/lifinity.ts [hunk 3]
```diff
@@ -243,21 +164,188 @@ const calculateMinimumTokenAmountFromLP = ({
...
+const getOutAmount = (
+  poolInfo: IPoolInfo,
+  amount: number | string,
+  fromCoinMint: string,
+  toCoinMint: string,
+  slippage: number,
+  coinBalance: BigNumber,
+  pcBalance: BigNumber,
+) => {
+  const price = pcBalance.dividedBy(coinBalance);
```
Comment (Orelsanpls) [inline tools/sdk/lifinity/lifinity.ts:176]: coinBalance can be 0?

File: tools/sdk/lifinity/lifinity.ts [hunk 4]
```diff
@@ -243,21 +164,188 @@ const calculateMinimumTokenAmountFromLP = ({
...
+  const fromAmount = new BigNumber(amount);
+  let outAmount = new BigNumber(0);
```
Comment (Orelsanpls) [inline tools/sdk/lifinity/lifinity.ts]: Alternative syntax proposition:

  ```
  const percent = new BigNumber(100)
    .plus(new BigNumber(slippage))
    .dividedBy(new BigNumber(100));

  if (!coinBalance || !pcBalance) {
    return new BigNumber(0);
  }

  if (
    fromCoinMint === poolInfo.poolCoinMint &&
    toCoinMint === poolInfo.poolPcMint
  ) {
    // outcoin is pc
    return fromAmount.multipliedBy(price).multipliedBy(percent);
  }
  
  if (
    fromCoinMint === poolInfo.poolPcMint &&
    toCoinMint === poolInfo.poolCoinMint
  ) {
    // outcoin is coin
    return fromAmount.dividedBy(percent).dividedBy(price);
  }
  
  return new BigNumber(0);
};
```

File: tools/sdk/lifinity/lifinity.ts [hunk 5]
```diff
@@ -243,21 +164,188 @@ const calculateMinimumTokenAmountFromLP = ({
...
+  const outAmount = getOutAmount(
+    pool,
+    amount.toString(),
+    coinAddress,
+    pcAddress,
+    slippage,
+    coinBalance,
+    pcBalance,
+  );
+  // Bruh
```
Comment (Orelsanpls) [inline tools/sdk/lifinity/lifinity.ts:252]: Bruh

File: tools/sdk/lifinity/lifinity.ts [hunk 6]
```diff
@@ -243,21 +164,188 @@ const calculateMinimumTokenAmountFromLP = ({
...
+  const lpRecive =
```
Comment (Orelsanpls) [inline tools/sdk/lifinity/lifinity.ts]: lpRecive -> lpReceived ?

File: tools/sdk/lifinity/lifinity.ts [hunk 7]
```diff
@@ -243,21 +164,188 @@ const calculateMinimumTokenAmountFromLP = ({
...
+    Math.floor(
+      ((amount.toNumber() * Math.pow(10, pool.poolCoinDecimal)) /
+        coinBalance.toNumber()) *
+        lpSupply.toNumber(),
+    ) / Math.pow(10, pool.poolMintDecimal);
+  const amountOut =
+    Math.floor(outAmount.toNumber() * Math.pow(10, pool.poolPcDecimal)) /
```
Comment (GregoryNEUT) [inline tools/sdk/lifinity/lifinity.ts:256]: Is there not an utils for that?

**Review by Orelsanpls — CHANGES_REQUESTED** [review]

**Review by Orelsanpls — APPROVED** [review]
Talked with @ThomasProust directly

**Pull request merged**